### PR TITLE
Added 2 missing testcases for Keras switch backend

### PR DESCRIPTION
**Pull request by amitsrivastava78**
There was no test cases for the keras backend switch added the same

### CODE REVIEW

File: tensorflow/python/keras/backend_test.py [hunk 1]
```diff
@@ -1736,5 +1736,40 @@ def test_batch_normalization(self):
...
+                               dtype='float32')
+    x1 = keras.backend.square(x)
+    def true_fun():
+      output = x * x
+      return output
+
+    def false_fun():
+      return y1
+
+    x2 = keras.backend.switch(keras.backend.equal(x, y1), true_fun, false_fun)
```
Comment (fchollet) [inline tensorflow/python/keras/backend_test.py]: You could simply this code considerably by having:

condition: `array_ops.constant(True)` or `array_ops.constant(False)` (or a boolean placeholder in order to test dynamic behavior)
And for one of the functions, you can use for example: `true_fn=lambda x: array_ops.constant(True)` (easier to check the output)

File: tensorflow/python/keras/backend_test.py [hunk 2]
```diff
@@ -1736,5 +1736,40 @@ def test_batch_normalization(self):
...
+    self.assertAllClose(x2, x1)
+
+    y2 = keras.backend.switch(keras.backend.equal(x, x2), true_fun, false_fun)
+    self.assertAllClose(y2, y1)
```
Comment (fchollet) [inline tensorflow/python/keras/backend_test.py]: I believe you should be running the graph (e.g. via `backend.function()`) in order to test the behavior of `switch`

**Review by fchollet — CHANGES_REQUESTED** [review]
Thanks for the PR!

File: tensorflow/python/keras/backend_test.py [hunk 3]
```diff
@@ -1736,5 +1736,40 @@ def test_batch_normalization(self):
     self.assertEqual(normed.shape.as_list(), [10, 3, 5, 5])
 
 
+class ControlOpsTests(test.TestCase):
+  @test_util.run_in_graph_and_eager_modes
```
[the remark this answers is not in the document]
Comment (amitsrivastava78) [reply tensorflow/python/keras/backend_test.py]: Updated as per comments

File: tensorflow/python/keras/backend_test.py (hunk 2 above)
Comment (amitsrivastava78) [reply tensorflow/python/keras/backend_test.py]: Updated as per comments

**Comment by amitsrivastava78** [comment]
@fchollet , thanks for the review, i have updated the code as per the review comments, kindly check and merge.

**Comment by amitsrivastava78** [comment]
@fchollet & @hgadig , can you pls review the PR

**Comment by hgadig** [comment]
> @fchollet & @hgadig , can you pls review the PR

@fchollet  PTAL

**Comment by amitsrivastava78** [comment]
@fchollet & @hgadig , all the code is updated as per your comments, can you pls review the PR

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @fchollet: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 14 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @fchollet: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 29 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @fchollet: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 44 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @fchollet, @fchollet: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 14 days with no activity and the `awaiting review` label has been applied.

File: tensorflow/python/keras/backend_test.py [hunk 4]
```diff
@@ -1736,5 +1736,34 @@ def test_batch_normalization(self):
     self.assertEqual(normed.shape.as_list(), [10, 3, 5, 5])
 
 
+class ControlOpsTests(test.TestCase):
+
+  def test_function_switch_basics(self):
```
Comment (qlzh727) [inline tensorflow/python/keras/backend_test.py]: Can we test this under both graph and eager mode? eg add @test_util.run_all_in_graph_and_eager_modes to the test class?

File: tensorflow/python/keras/backend_test.py [hunk 5]
```diff
@@ -1736,5 +1736,34 @@ def test_batch_normalization(self):
...
+    pred = keras.backend.placeholder([], dtype='bool')
+
+    def false_fun():
+      return array_ops.constant(False)
+
+    with self.cached_session():
+      x2 = keras.backend.switch(pred, lambda: array_ops.constant(True), false_fun)
```
Comment (qlzh727) [inline tensorflow/python/keras/backend_test.py]: Just curious why the true branch is a lambda, while the false branch is a function, can we use lambda for both case for simplicity?

File: tensorflow/python/keras/backend_test.py [hunk 6]
```diff
@@ -1736,5 +1736,34 @@ def test_batch_normalization(self):
...
+      x1 = keras.backend.function(pred, x2)([True])
```
Comment (qlzh727) [inline tensorflow/python/keras/backend_test.py]: The x1/x2 naming here is bit unclear to me, x2 here is a tensor and x1 is a value. Can we give them proper name?

File: tensorflow/python/keras/backend_test.py [hunk 7]
```diff
@@ -1736,5 +1736,34 @@ def test_batch_normalization(self):
...
+      self.assertAllClose(True, x1)
```
Comment (qlzh727) [inline tensorflow/python/keras/backend_test.py]: Since they are just simple value, we can just use assertEqual().

File: tensorflow/python/keras/backend_test.py [hunk 8]
```diff
@@ -1736,5 +1736,34 @@ def test_batch_normalization(self):
...
+    x = ops.convert_to_tensor(np.array([[1, 2, 3], [4, 5, 6]]),
+                              dtype='float32')
+    y1 = ops.convert_to_tensor(np.array([1, 2, 3]), dtype='float32')
+    def true_fun():
+      return x
+
+    def false_fun():
+      return y1
+
+    with self.assertRaises(ValueError):
```
Comment (qlzh727) [inline tensorflow/python/keras/backend_test.py]: Just by looking at the test, it is not very clear to me why the error is thrown. When I look at the code, I noticed that the error is caused by different rank of input/output. Can we rename the test case to include that info, and also update the assertion here by self.assertRaisesRegexp so that the key part of the error message is also included.

File: tensorflow/python/keras/backend_test.py [hunk 9]
```diff
@@ -1736,5 +1736,34 @@ def test_batch_normalization(self):
...
+      x1 = keras.backend.function(pred, x2)([True])
+      self.assertAllClose(True, x1)
+      x1 = keras.backend.function(pred, x2)([False])
+      self.assertAllClose(False, x1)
+
+  def test_function_basic_error(self):
+    x = ops.convert_to_tensor(np.array([[1, 2, 3], [4, 5, 6]]),
+                              dtype='float32')
+    y1 = ops.convert_to_tensor(np.array([1, 2, 3]), dtype='float32')
+    def true_fun():
```
Comment (qlzh727) [inline tensorflow/python/keras/backend_test.py]: let rename this to true_func.

File: tensorflow/python/keras/backend_test.py [hunk 10]
```diff
@@ -1736,5 +1736,34 @@ def test_batch_normalization(self):
...
+      return x
+
+    def false_fun():
```
Comment (qlzh727) [inline tensorflow/python/keras/backend_test.py]: same here, false_func

File: tensorflow/python/keras/backend_test.py [hunk 11]
```diff
@@ -1736,5 +1736,34 @@ def test_batch_normalization(self):
...
+      x2 = keras.backend.switch(pred, lambda: array_ops.constant(True), false_fun)
+      x1 = keras.backend.function(pred, x2)([True])
+      self.assertAllClose(True, x1)
+      x1 = keras.backend.function(pred, x2)([False])
+      self.assertAllClose(False, x1)
+
+  def test_function_basic_error(self):
+    x = ops.convert_to_tensor(np.array([[1, 2, 3], [4, 5, 6]]),
+                              dtype='float32')
+    y1 = ops.convert_to_tensor(np.array([1, 2, 3]), dtype='float32')
```
Comment (qlzh727) [inline tensorflow/python/keras/backend_test.py]: let's call this y, since there isn't a y2 here.

**Review by qlzh727 — CHANGES_REQUESTED** [review]
Thanks for the change. Most of the comment are about variable naming and code style. Thanks.

File: tensorflow/python/keras/backend_test.py (hunk 5 above)
Comment (amitsrivastava78) [reply tensorflow/python/keras/backend_test.py]: Sure, same is updated in the code.

Regards
Amit

File: tensorflow/python/keras/backend_test.py (hunk 7 above)
Comment (amitsrivastava78) [reply tensorflow/python/keras/backend_test.py]: Sure, same is updated in the code.

Regards
Amit

File: tensorflow/python/keras/backend_test.py (hunk 8 above)
Comment (amitsrivastava78) [reply tensorflow/python/keras/backend_test.py]: Sure, same is updated in the code.

Regards
Amit

File: tensorflow/python/keras/backend_test.py (hunk 10 above)
Comment (amitsrivastava78) [reply tensorflow/python/keras/backend_test.py]: Sure, same is updated in the code.

Regards
Amit

File: tensorflow/python/keras/backend_test.py (hunk 11 above)
Comment (amitsrivastava78) [reply tensorflow/python/keras/backend_test.py]: Sure, same is updated in the code.

Regards
Amit

File: tensorflow/python/keras/backend_test.py (hunk 9 above)
Comment (amitsrivastava78) [reply tensorflow/python/keras/backend_test.py]: Sure, same is updated in the code.

Regards
Amit

File: tensorflow/python/keras/backend_test.py (hunk 6 above)
Comment (amitsrivastava78) [reply tensorflow/python/keras/backend_test.py]: Sure, same is updated in the code.

Regards
Amit

File: tensorflow/python/keras/backend_test.py (hunk 4 above)
Comment (amitsrivastava78) [reply tensorflow/python/keras/backend_test.py]: Sure, the same has been updated in the code.

Regards
Amit

Comment (qlzh727) [reply tensorflow/python/keras/backend_test.py]: The annotation should be applied to the test class, which will cover all the test case within it.

File: tensorflow/python/keras/backend_test.py [hunk 12]
```diff
@@ -1880,5 +1880,33 @@ def test_get_session_different_graphs(self):
...
+class ControlOpsTests(test.TestCase):
+
+  @test_util.run_all_in_graph_and_eager_modes
+  def test_function_switch_basics(self):
+    pred = keras.backend.placeholder([], dtype='bool')
+
+    tensor = keras.backend.switch(pred, lambda: array_ops.constant(True),
+                                  lambda: array_ops.constant(False))
+    value = keras.backend.function(pred, tensor)([True])
+    self.assertEqual(True, value)
```
Comment (qlzh727) [inline tensorflow/python/keras/backend_test.py]: This can be replaced with self.assertTrue(value) which is more readable.

File: tensorflow/python/keras/backend_test.py [hunk 13]
```diff
@@ -1880,5 +1880,33 @@ def test_get_session_different_graphs(self):
...
+    value = keras.backend.function(pred, tensor)([False])
+    self.assertEqual(False, value)
```
Comment (qlzh727) [inline tensorflow/python/keras/backend_test.py]: same here, self.assertFalse().

File: tensorflow/python/keras/backend_test.py [hunk 14]
```diff
@@ -1880,5 +1880,33 @@ def test_get_session_different_graphs(self):
...
+
+  @test_util.run_all_in_graph_and_eager_modes
```
Comment (qlzh727) [inline tensorflow/python/keras/backend_test.py]: Please remove this annotation, and move it to the class level.

**Review by qlzh727 — CHANGES_REQUESTED** [review]

File: tensorflow/python/keras/backend_test.py (hunk 12 above)
Comment (amitsrivastava78) [reply tensorflow/python/keras/backend_test.py]: Sure, this is done as suggested by you.

Regards
Amit

File: tensorflow/python/keras/backend_test.py (hunk 13 above)
Comment (amitsrivastava78) [reply tensorflow/python/keras/backend_test.py]: Sure, done as per your suggestion.

Regards
Amit

File: tensorflow/python/keras/backend_test.py (hunk 14 above)
Comment (amitsrivastava78) [reply tensorflow/python/keras/backend_test.py]: eager mode execution with bool has issues, i guess bool does not seems to be working in eager mode. So i have removed the annotation.

Regards
Amit

Comment (qlzh727) [reply tensorflow/python/keras/backend_test.py]: I think it is important to test both eager and graph mode, and eager mode default on in tf 2.0. Please add the annotation ,and we can check why it is failing.

Comment (amitsrivastava78) [reply tensorflow/python/keras/backend_test.py]: > I think it is important to test both eager and graph mode, and eager mode default on in tf 2.0. Please add the annotation ,and we can check why it is failing.

@qlzh727 , thanks alot, i will try to fix this by tomorrow and if i can not then will seek your help, thanks alot for reaching out like this.

Regards
Amit

**Comment by amitsrivastava78** [comment]
@qlzh727 , i have updated the code as per your suggestion, kindly review.

Regards
Amit

**Comment by amitsrivastava78** [comment]
@qlzh727 , i have updated the code as per your suggestion, kindly review.

Regards
Amit

**Review by qlzh727 — APPROVED** [review]
